### PR TITLE
Make the iter_kv_map lint handle ref/mut annotations.

### PR DESCRIPTION
**Pull request by khuey**
For the degenerate (`map(|(k, _)| k)`/`map(|(_, v)| v)`) cases a mut annotation is superfluous and a ref annotation won't compile, so no additional handling is required. For cases where the `map` call must be preserved ref/mut annotations should also be presereved so that the map body continues to work as expected.

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog: [`iter_kv_map`]: handle ref/mut annotations

### CODE REVIEW

File: clippy_lints/src/methods/iter_kv_map.rs [hunk 1]
```diff
@@ -30,9 +30,9 @@ pub(super) fn check<'tcx>(
         if let Body {params: [p], value: body_expr, generator_kind: _ } = cx.tcx.hir().body(c.body);
         if let PatKind::Tuple([key_pat, val_pat], _) = p.pat.kind;
 
-        let (replacement_kind, binded_ident) = match (&key_pat.kind, &val_pat.kind) {
-            (key, PatKind::Binding(_, _, value, _)) if pat_is_wild(cx, key, m_arg) => ("value", value),
-            (PatKind::Binding(_, _, key, _), value) if pat_is_wild(cx, value, m_arg) => ("key", key),
+        let (replacement_kind, annotation, binded_ident) = match (&key_pat.kind, &val_pat.kind) {
```
Comment (llogiq) [inline clippy_lints/src/methods/iter_kv_map.rs:33]: `binded_ident` is not correct English. Please rename to `bound_ident`.

**Review by llogiq — APPROVED** [review]
Nice! I just have one very minor niggle, otherwise this looks good to merge.

**Comment by khuey** [comment]
Done.

**Comment by llogiq** [comment]
Thank you!

@bors r+

**Pull request merged**